Added native systemd service

spacecowboy opened the pull request:
and moved the sysvinit service so it is installed by dh_installinit.

dh_installinit will do the right thing depending on the system you are
installing on. If the system has systemd, it will enable the native
systemd unit. If it has sysvinit, it will enable the old sysvinit
service.

I have also made sure that the service will be called `neo4j` for both
the neo4j and neo4j-enterprise packages.

I have tested locally (Debian Stretch) that the systemd service is installed correctly.

I tested on Ubuntu 14.04 (which does not have systemd) in VirtualBox and verified that it installs and enables the correct service there:

- [x] Make sure an existing `/etc/default/neo4j` is parsed on upgrade and put into `/etc/systemd/system/neo4j.service.d/override.conf`

spacecowboy committed:
Fix some build warnings

spacecowboy committed:
Added native systemd service

and moved the sysvinit service so it is installed by dh_installinit.

dh_installinit will do the right thing depending on the system you are
installing on. If the system has systemd, it will enable the native
systemd unit. If it has sysvinit, it will enable the old sysvinit
service.

I have also made sure that the service will be called `neo4j` for both
the neo4j and neo4j-enterprise packages.

benbc commented:
@spacecowboy This looks great. Thanks.

I think this will (or at least should!) involve documentation changes for the timeout and file handle limit settings.

I believe that the changes you've made so that "the service will be called neo4j for both
the neo4j and neo4j-enterprise packages" is effectively just an implementation detail of using `dh_installinit`. You haven't changed the name of the service from 3.1, have you?

The Quality team are expanding the matrix of compatibility builds. Please talk to them to ensure that there will be (at some point soon) two different distros/versions in tyre-kicking with the Debian package so that we hit both service definitions.

spacecowboy committed:
Change indentation of preinst file

spacecowboy committed:
Fix some line endings

spacecowboy committed:
Migrate users /etc/default/neo4j file to systemd overrides

Make sure to be backwards compatible by migrating the settings defined
by users in /etc/default/neo4j to the appropriate place with systemd.

Also change the defaults file to be entirely commented out as to not
trigger this behavior by default for future upgrades.

spacecowboy commented:
@benbc added code to handle migration of defaults file

You are correct in that the service name is not a change - it's just that we used to copy the service and defaults file manually and not let debian do it for us.

Examples of the upgrade process

## User has modified /etc/default/neo4j int their 3.1.0 install and is upgrading to 3.2.0

![selection_012](https://cloud.githubusercontent.com/assets/223655/21893883/450b5e8a-d8dd-11e6-8820-08d562d4cf15.png)

Migration is performed and then the user is asked by debian what to do with the changed defaults file

## User views diff between files

![selection_013](https://cloud.githubusercontent.com/assets/223655/21893904/5eb30626-d8dd-11e6-9a98-24f0ac1a02dd.png)

And can choose whatever he wants (we have already migrated the file)

## After install, resulting service file

![selection_014](https://cloud.githubusercontent.com/assets/223655/21893926/7c640d96-d8dd-11e6-9fd1-2d022bae0dd1.png)

benbc commented on new-packaging/src/debian/skeleton/debian/neo4j.default:
```diff
@@ -7,10 +7,11 @@
 #
 # Max number of open file descriptors
-NEO4J_ULIMIT_NOFILE=40000
+#NEO4J_ULIMIT_NOFILE=60000
```
Why this change?

spacecowboy replied:
Value: 
I thought I was being consistent with what we shipped in the release 3.1.0 package but looking at the [source](https://github.com/neo4j/neo4j/commit/1fc3c0ab9f352f883a60cb126d25496a6f61d9e6) it seems I was mistaken.

Commented:
Being commented out by default is to prevent an unnecessary migration to happen when you go from 3.2.0 to 3.2.0+, it was also commented out before, so in the final diff this is not a change.

benbc commented on new-packaging/src/debian/skeleton/debian/neo4j.default:
```diff
@@ -7,10 +7,11 @@
 #
 #     systemctl edit neo4j.service
 #
+# This is automatically done upon upgrade.
```
I think this would be more helpful if it was a bit more verbose. Something like:

```
# Any values that were present in this file have been automatically migrated into
# SystemD configuration as part of the upgrade.
```

benbc reviewed (commented):
Looks good other than these two things.

spacecowboy committed:
Improve comment about move to systemd

and change the NOFILE value to be the same as old defaults file.
